packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol

```diff
@@ -5,6 +5,40 @@ pragma solidity ^0.8.0;
 /// @notice Interface for implementing a leader election scheme
 
 abstract contract LeaderElectionBatchInbox {
+    enum LeaderStatusFlags {
+        Scheduled,
+        Unscheduled,
+        Invalid
+    }
+
+    struct MetaData {
+        /// Frame metadata
+        bytes16 channelId;
+        uint16 frameNumber;
+        uint32 frameDataLength;
+        bool isLast;
+        /// The number of L2 blocks in this channel (including the current frame).
+        uint16 numL2Blocks;
+    }
+
+    /// @notice Allows to submit a batch. This function checks that the caller is the leader for the current block.
+    /// @notice The caller of this function is currently expected to append the frames as calldata after
+    ///         the metadata. This is done to avoid copying the frames in memory and allows the use of a custom
+    ///         encoding for the frames, without having to RLP-decode the frames.
+    ///
+    ///         Checking that the metadata matches the frames must be done in the derivation pipeline. If there is
+    ///         a mismatch between the metadata and the frames, the frames must be discarded.
+    ///
+    ///         Implementations of this contract interface can enforce behaviour based on the metadata submitted
+    ///         to this function.
+    ///
+    /// @param _metas metadata for the frames.
+    function submit(MetaData[] memory _metas) public view {
```

Comment:
This should be `external`, not `public`, and absolutely not `view` (I removed the `view` in [PR 48](https://github.com/EspressoSystems/op-leader-election/pull/48)).

 We're never going to call `submit` from inside another contract.
 
 And then we can change the signature to 
 ```solidity
 function submit(MetaData[] memory _metas, bytes calldata _batch) external {
 ```
 which will not copy the calldata into memory.
 
 It may be possible to keep the metadata as calldata and copy it, bytewise, into the series of needed locals, but I'm not sure if this saves us anything.